### DISCUSSION TITLE
CGI

### DISCUSSION THREAD

**Discussion by anolivei** [42vogons/webserv#13] [General]
1.  Execute CGI based on certain file extension (for example .php).
2.  Make it work with POST and GET methods.
3. Make the route able to accept uploaded files and configure where they should be saved.
- Do you wonder what a CGI is?
-  Because you won’t call the CGI directly, use the full path as PATH_INFO.
- Just remember that, for chunked request, your server needs to unchunk it, the CGI will expect EOF as end of the body.
- Same things for the output of the CGI. If no content_length is returned from the CGI, EOF will mark the end of the returned data.
- Your program should call the CGI with the file requested as first argument.
- The CGI should be run in the correct directory for relative path file access.
- Your server should work with one CGI (php-CGI, Python, and so forth).

**Comment by anolivei** [comment]
> 1. Execute CGI based on certain file extension (for example .php).
> 2. Make it work with POST and GET methods.
> 3. Make the route able to accept uploaded files and configure where they should be saved.
> 
> * Do you wonder what a CGI is?
> * Because you won’t call the CGI directly, use the full path as PATH_INFO.
> * Just remember that, for chunked request, your server needs to unchunk it, the CGI will expect EOF as end of the body.
> * Same things for the output of the CGI. If no content_length is returned from the CGI, EOF will mark the end of the returned data.
> * Your program should call the CGI with the file requested as first argument.
> * The CGI should be run in the correct directory for relative path file access.
> * Your server should work with one CGI (php-CGI, Python, and so forth).

https://en.wikipedia.org/wiki/Common_Gateway_Interface

**Comment by cpereira42** [comment]
```go
package main

import (
	"fmt"
	"net/http"
	"os"
)

func main() {
	// Obtém o valor da variável de ambiente "QUERY_STRING"
	queryString := os.Getenv("QUERY_STRING")

	// Define o tipo de conteúdo da resposta como texto simples
	fmt.Println("Content-Type: text/plain")

	// Escreve o cabeçalho em branco para indicar o fim do cabeçalho HTTP
	fmt.Println()

	// Escreve a consulta de entrada na resposta
	fmt.Printf("Query string: %s\n", queryString)

	// Define o código de status HTTP como 200 (OK)
	statusCode := http.StatusOK

	// Define o código de status HTTP na resposta
	fmt.Printf("Status: %d %s\n", statusCode, http.StatusText(statusCode))
}
```
